presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +494,48 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        if (preGroupedHashStrategy.isPresent()) {
+            // Running in segmented aggregation mode.
+            int lastRowInPage = page.getPositionCount() - 1;
+            if (firstUnfinishedSegment == null) {
+                // If this is the first page, treat the first segment in this page as the current segment.
+                firstUnfinishedSegment = page.getRegion(0, 1);
+            }
+
+            if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), lastRowInPage, page.extractChannels(preGroupedChannels))) {
+                // If the current segment might have more data in the incoming pages, process the whole page.
+                unfinishedWork = aggregationBuilder.processPage(page);
+            }
+            else if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), 0, page.extractChannels(preGroupedChannels))) {
+                // If the current page starts with a new segment, flush before processing it.
+                remainingPageForSegmentedAggregation = page;
+            }
+            else {
+                // If the current segment ends in the current page, flush it with all the segments (if exist) except the last segment of the current page.
+                int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), page.extractChannels(preGroupedChannels));
+                unfinishedWork = aggregationBuilder.processPage(page.getRegion(0, lastSegmentStart));
+                remainingPageForSegmentedAggregation = page.getRegion(lastSegmentStart, lastRowInPage - lastSegmentStart + 1);
+            }
+            // Record the last segment.
+            firstUnfinishedSegment = page.getRegion(lastRowInPage, 1);
+        }
+        else {
+            unfinishedWork = aggregationBuilder.processPage(page);
+        }
+    }
+
+    private int findLastSegmentStart(PagesHashStrategy pagesHashStrategy, Page page)
+    {
+        for (int i = page.getPositionCount() - 1; i > 0; i--) {
```

Comment:
Will binary search improve the performance?

Reply:
what do you think about this one? @zacw7

Reply:
Implemented binary search (see the most recent code in this PR) and have it tested 20220621_220634_00011_rzc9n (vs linear search 20220615_223659_00013_cnhk9). Don't see much difference, do we want to keep the change?